core/base/src/TSystem.cxx

```diff
@@ -3782,18 +3782,37 @@ int TSystem::CompileMacro(const char *filename, Option_t *opt,
          ::Info("ACLiC","compiling the dictionary and script files");
          if (verboseLevel>4)  ::Info("ACLiC", "%s", cmd.Data());
       }
-      Int_t compilationResult = gSystem->Exec( cmd );
-      if (compilationResult) {
-         if (compilationResult==139) ::Error("ACLiC","Compilation failed with a core dump!");
-         else ::Error("ACLiC","Compilation failed!");
+      Int_t success = ExecAndReport( cmd );
+      if (!success) {
          if (produceRootmap) {
             gSystem->Unlink(libmapfilename);
          }
       }
-      result = !compilationResult;
+      result = success;
    }
 
    if ( result ) {
+      if (linkDepLibraries) {
+         // We may have unresolved symbols. Use dyld to resolve the dependent
+         // libraries and relink.
+         // FIXME: We will likely have duplicated libraries as we are appending
+         // FIXME: This likely makes rootcling --lib-list-prefix redundant.
+         TString depLibsFullPaths;
+         std::function<bool(const char*)> CollectF = [&depLibsFullPaths](const char* dep) {
+            TString LibFullPath(dep);
+            if (!gSystem->FindDynamicLibrary(LibFullPath, /*quiet=*/true)) {
+               ::Error("TSystem::CompileMacro", "Cannot find library '%s'", dep);
+               return false; // abort
+            }
+            depLibsFullPaths += " " + LibFullPath;
+            return true;
+         };
+         ForeachSharedLibDep(library, CollectF);
+
+         TString relink_cmd = cmd.Strip(TString::kTrailing, ';');
```

Comment:
@pcanal, this is what I came up with. So, we create the shared library, then we check if we have explicit linking on and if we do we ask the dynamic linker for the actual library dependencies and relink the library.

This might allow us to not depend on `rootcling --lib-list-prefix` at all.

Reply:
Indeed, it seems it might work well.